packages/root/src/Root.tsx

```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { HTMLElementProps } from '@leafygreen-ui/lib';
+
+type ButtonProps<T> = HTMLElementProps<'button'> &
+  T & {
+    href?: undefined;
+  };
+
+type Anchor<T> = HTMLElementProps<'a'> &
+  T & {
+    href: string;
+  };
+
+type CustomElementProps<T> = T & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type RootProps<T> = ButtonProps<T> | Anchor<T> | CustomElementProps<T>;
+
+function usesCustomElement<T>(
+  props: RootProps<T>,
+): props is CustomElementProps<T> {
+  return (props as any).as != null;
+}
+
+function usesLinkElement<T>(props: RootProps<T>): props is Anchor<T> {
```

Comment:
[nitpick] super nitpicky, can we change this to `usesAnchorElement` because there is already a `Link` element (used in `<head>` though, not `<body>`)>?

Reply:
👍